databricks/koalas/frame.py

```diff
@@ -3062,10 +3062,23 @@ def between_time(
         if not isinstance(self.index, DatetimeIndex):
             raise TypeError("Index must be DatetimeIndex")
 
-        def pandas_between_time(pdf):
-            return pdf.between_time(start_time, end_time, include_start, include_end)
+        def pandas_between_time(
+            pdf,
+        ) -> ks.DataFrame[  # type: ignore
+            zip(self.reset_index().columns, self.reset_index().dtypes)
+        ]:
+            return pdf.between_time(start_time, end_time, include_start, include_end).reset_index()
+
+        # apply_batch will remove the index of the Koalas DataFrame and attach a default index,
+        # which will never be used. So use "distributed" index as a dummy to avoid overhead.
+        with option_context("compute.default_index_type", "distributed"):
+            # In pandas reset_index(), the new column is assigned the name 'index'
+            # if the original index has no name
+            index_name = "index" if self.index.name is None else self.index.name
```

Comment:
I'm afraid this is not good enough. 

```py
>>> kdf = ks.DataFrame({('X', 'A'): [1, 2, 3, 4]}, index=idx)
>>> kdf.between_time('0:15', '0:45')
Traceback (most recent call last):
...
TypeError: not all arguments converted during string formatting
```
or
```py
>>> kdf = ks.DataFrame({'index': [1, 2, 3, 4]}, index=idx)
>>> kdf.between_time('0:15', '0:45')
              level_0
2 2018-04-10 00:20:00
3 2018-04-11 00:40:00
```

Reply:
@xinrong-databricks Could you add the tests with the cases I posted above and the following?

```py
>>> idx = pd.date_range('2018-04-09', periods=4, freq='1D20min', name='index')
>>> kdf = ks.DataFrame({'index': [1, 2, 3, 4]}, index=idx)
>>> kdf.between_time('0:15', '0:45')
Traceback (most recent call last):
...
ValueError: cannot insert index, already exists
```

Reply:
Sounds good!